### PR TITLE
feat: add success and error callbacks to BulkWriter

### PR DESCRIPTION
**Pull request by thebrianchen**
Continuation of #458 but without blocking gax/grpc executor!

### CODE REVIEW

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 1]
```diff
@@ -201,35 +300,44 @@
...
 
   /**
    * Write to the document referred to by the provided DocumentReference. If the document does not
    * exist yet, it will be created.
    *
    * @param documentReference A reference to the document to be set.
-   * @param fields A map of the fields and values for the document.
-   * @return An ApiFuture containing a sentinel value (Timestamp(0)) for the delete operation.
-   *     Contains an error if the delete fails.
+   * @param fields A map of the fields and values for the document. * @return An ApiFuture
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:325]: Please fix "return" part.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 2]
```diff
@@ -262,16 +375,23 @@
...
    * @param options An object to configure the set behavior.
-   * @return An ApiFuture containing the result of the write. Contains an error if the write fails.
+   * @return An ApiFuture containing the result of the write. Contains a {@link BulkWriterError} if
+   *     the write fails.
    */
   @Nonnull
   public ApiFuture<WriteResult> set(
-      @Nonnull DocumentReference documentReference, Object pojo, @Nonnull SetOptions options) {
+      @Nonnull final DocumentReference documentReference,
+      final Object pojo,
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:384]: Drive-by comment: Should this be Nonnull as well?

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 3]
```diff
@@ -442,22 +591,163 @@
...
+            MoreExecutors.directExecutor());
+
+    writeResultApiFuture.addListener(
+        new Runnable() {
+          public void run() {
+            pendingOperations.remove(operationCompletedFuture);
+            operationCompletedFuture.set(null);
+          }
+        },
+        MoreExecutors.directExecutor());
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:646]: I think this is where you want to use your own Executor to remove the need to synchronization on the internal data structures.  To make this easier to parse, I would also suggest doing the same in line 637 (just above)

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 4]
```diff
@@ -442,22 +591,163 @@
...
+                    exception.getStatus().getCode().value(), shouldRetry));
+            if (!shouldRetry) {
+              throw bulkWriterError;
+            } else {
+              return enqueueWriteHelper(
+                  documentReference, operationType, operationCallback, failedAttempts + 1);
+            }
+          }
+        },
+        MoreExecutors.directExecutor());
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:713]: Again, it would be better to more explicit here and specify the executor.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 5]
```diff
@@ -442,22 +591,163 @@
...
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback,
+      final int failedAttempts) {
+    List<BulkCommitBatch> operationBatchQueue = failedAttempts > 0 ? retryBatchQueue : batchQueue;
+    BulkCommitBatch bulkCommitBatch = getEligibleBatch(documentReference, operationBatchQueue);
+
+    // Send ready batches if this is the first attempt. Subsequent retry batches are scheduled
+    // after the initial batch returns.
+    if (failedAttempts == 0) {
+      sendReadyBatches(operationBatchQueue);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:682]: Can we move this to `enqueueWrite`? I think this would allow you to get rid of this if-check.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 6]
```diff
@@ -442,22 +591,163 @@
...
+                    exception.getMessage(),
+                    documentReference,
+                    operationType,
+                    failedAttempts);
+
+            boolean shouldRetry = invokeUserErrorCallback(bulkWriterError).get();
+            logger.log(
+                Level.INFO,
+                String.format(
+                    "Running error callback on error code: %d, shouldRetry: %b",
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:703]: I think this log message would be more useful if it included the DocumentReference. It is also printed after the callback ran - so ideally, we would rephrase it a bit.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 7]
```diff
@@ -442,22 +591,163 @@
...
+            }
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  /** Invokes the user error callback on the user callback executor and returns the result. */
+  private SettableApiFuture<Boolean> invokeUserErrorCallback(final BulkWriterError error) {
+    final SettableApiFuture<Boolean> callbackResult = SettableApiFuture.create();
+    errorExecutor.execute(
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:719]: If you make the default Exectuor the `bulkWriterExecutor`, it might be best to make the `errorExecutor` nullable. If no custom executor is set, you could invoke the `errorListener` directly.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 8]
```diff
@@ -442,22 +591,163 @@
...
+            }
+          }
+        });
+    return callbackResult;
+  }
+
+  /** Invokes the user success callback on the user callback executor. */
+  private ApiFuture<Void> invokeUserSuccessCallback(
+      final DocumentReference documentReference, final WriteResult result) {
+    final SettableApiFuture<Void> callbackResult = SettableApiFuture.create();
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:737]: Same comment as above.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 9]
```diff
@@ -477,28 +767,66 @@
...
               }
             },
             MoreExecutors.directExecutor());
+
+    // Afterwards, send all accumulated retry operations. Wait until the retryBatchQueue is cleared.
+    // This way, operations scheduled after flush() will not be sent until the retries are done.
+    batchQueueFlushComplete.addListener(
+        new Runnable() {
+          public void run() {
+            if (retryBatchQueue.size() > 0) {
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:805]: Is this if-check needed?

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 10]
```diff
@@ -477,28 +767,66 @@
...
+              for (BulkCommitBatch batch : retryBatchQueue) {
+                batch.markReadyToSend();
+              }
+              sendReadyBatches(retryBatchQueue);
+            }
+            ApiFutures.successfulAsList(pendingOperations)
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:811]: I think this doesn't need to be chained here. Could we move it outside of this block to reduce chaining a bit?

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 11]
```diff
@@ -515,70 +843,173 @@ private void verifyNotClosed() {
...
 
       // Send the batch if it is under the rate limit, or schedule another attempt after the
       // appropriate timeout.
       long delayMs = rateLimiter.getNextRequestDelayMs(batch.getPendingOperationCount());
       Preconditions.checkState(delayMs != -1, "Batch size should be under capacity");
       if (delayMs == 0) {
-        sendBatch(batch);
+        sendBatch(batch, batchQueue, batchCompletedFuture);
       } else {
         firestoreExecutor.schedule(
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:1006]: Can we use our new executor here as well? It would need to be a `ScheduledExecutor`.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterError.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  public Status getStatus() {
+    return status;
+  }
+
+  /** @return The error message of the error. */
+  public String getMessage() {
+    return message;
+  }
+
+  /** @return The document reference the operation was performed on. */
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterError.java:34]: "A reference to the document" or "The DocumentReference"

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java [hunk 13]
```diff
@@ -680,169 +633,18 @@ private T performDelete(
...
-    return false;
-  }
-
-  void markReadyToSend() {
-    if (state == BatchState.OPEN) {
-      state = BatchState.READY_TO_SEND;
-    }
+  /** Get the number of writes. */
+  @VisibleForTesting
+  int getMutationsSize() {
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java:647]: We should probably settle on either "write" or "mutation" in this comment and method name. I think it should also be singular: "getWriteCount()".

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+    BulkWriter writer = firestore.bulkWriter();
+    ApiFuture<WriteResult> result = writer.update(docRef, "foo", "newValue");
+    writer.close();
+
+    try {
+      result.get();
+      fail("Update operation should have thrown exception");
+    } catch (Exception e) {
+      assertTrue(e.getMessage().contains("No document to update"));
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java:139]: Does the emulator now print this error message? It used to print "No entity to update". It would be nice if these tests passed against Prod and the emulator.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    }
+    Executor executor = Executors.newSingleThreadExecutor(new NamedThreadFactory());
+    final List<String> operations = new ArrayList<>();
+
+    BulkWriter writer = firestore.bulkWriter();
+    writer.addWriteResultListener(
+        executor,
+        new WriteResultCallback() {
+          public void onResult(DocumentReference documentReference, WriteResult result) {
+            operations.add("operation");
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java:174]: Do you want to verify the Thread name here?

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+          }
+        });
+    writer.set(randomDoc, Collections.singletonMap("foo", "bar"));
+    writer.flush().get();
+    assertEquals("operation", operations.get(0));
+  }
+}
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java:181]: We should also test the error handler (with a failed precondition for example).

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 17]
```diff
@@ -352,6 +351,328 @@ public void sendWritesToDifferentDocsInSameBatch() throws Exception {
...
+        new WriteErrorCallback() {
+          public boolean onError(BulkWriterError error) {
+            errorListenerCalled[0] = true;
+            assertEquals(Status.INTERNAL, error.getStatus());
+            return false;
+          }
+        });
+
+    ApiFuture<WriteResult> result1 = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close();
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:469]: s/result1/result

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 18]
```diff
@@ -352,6 +351,328 @@ public void sendWritesToDifferentDocsInSameBatch() throws Exception {
...
+
+    bulkWriter.addWriteErrorListener(
+        new WriteErrorCallback() {
+          public boolean onError(BulkWriterError error) {
+            throw new UnsupportedOperationException(
+                "Test code threw UnsupportedOperationException");
+          }
+        });
+
+    ApiFuture<WriteResult> result1 = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:499]: s/result1/result

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 19]
```diff
@@ -352,6 +351,328 @@ public void sendWritesToDifferentDocsInSameBatch() throws Exception {
...
+
+    bulkWriter.addWriteResultListener(
+        new WriteResultCallback() {
+          public void onResult(DocumentReference documentReference, WriteResult result) {
+            throw new UnsupportedOperationException(
+                "Test code threw UnsupportedOperationException");
+          }
+        });
+
+    ApiFuture<WriteResult> result1 = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:529]: s/result1/result

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 20]
```diff
@@ -352,6 +351,328 @@ public void sendWritesToDifferentDocsInSameBatch() throws Exception {
...
+              public void run() {
+                operations.add("AFTER_FLUSH");
+              }
+            },
+            executor);
+    flushComplete.get();
+
+    // Verify that the 2nd operation did not complete as a result of the flush() call.
+    assertArrayEquals(new String[] {"BEFORE_FLUSH", "FLUSH"}, operations.toArray());
+    bulkWriter.close();
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:635]: It might make sense to add another flush here and verify "AFTER_FLUSH" as well. That way, it doesn't look unused.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 21]
```diff
@@ -422,8 +743,8 @@ public void retriesIndividualWritesThatFailWithAbortedOrUnavailable() throws Exc
       result1.get();
       fail("set() should have failed");
     } catch (Exception e) {
-      assertTrue(e.getCause() instanceof FirestoreException);
-      assertEquals(Status.DEADLINE_EXCEEDED, ((FirestoreException) e.getCause()).getStatus());
+      assertTrue(e.getCause() instanceof BulkWriterError);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:746]: Could `BulkWriterError` extend `FirestoreException`? That way, it can be commonly treated by our users.

`BulkWriterError` should likely also be `BulkWriterException`. See the first part of the first answer here:  https://stackoverflow.com/questions/5813614/what-is-difference-between-errors-and-exceptions

**Review by schmidt-sebastian — COMMENTED** [review]
Very nice. I am happy that we managed to get this to work!

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java (hunk 14 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java]: Ah, it still prints "entity". Added the regex matcher instead.

Just wondering, do we have any plans to run integration tests against the emulator for server SDKs?

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java (hunk 16 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java]: Done.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java (hunk 21 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java]: I attempted it. I had to make `FirestoreException` not `final` in order to extend, and I had to make one of the private constructors package-private.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkCommitBatch.java [hunk 22]
```diff
@@ -16,32 +16,177 @@
...
+    request.setDatabase(firestore.getDatabaseName());
+
+    for (WriteOperation writeOperation : getWrites()) {
+      request.addWrites(writeOperation.write);
     }
 
+    ApiFuture<BatchWriteResponse> response =
+        firestore.sendRequest(request.build(), firestore.getClient().batchWriteCallable());
+
+    committed = true;
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkCommitBatch.java]: Done. I'll back-port this to node later along with some of the other changes like the retry logging message.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java (hunk 14 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterIntegrationTest.java]: Not yet I think. It helps during development of new features though, so we make an effort to ensure that new tests run against both environments.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 23]
```diff
@@ -82,28 +117,77 @@
    * modification errors (as this list is modified from both the user thread and the network
    * thread).
    */
-  private final List<BulkCommitBatch> batchQueue = new CopyOnWriteArrayList<>();
+  private final List<BulkCommitBatch> batchQueue = new ArrayList<>();
+
+  /**
+   * A queue of batches to be retried. Use a synchronized list to avoid multi-thread concurrent
+   * modification errors (as this list is modified from both the user thread and the network
+   * thread).
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:125]: Comment is outdated.

**Review by schmidt-sebastian — APPROVED** [review]
LGTM. Thanks for not giving up!

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 23 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java]: updated.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 24]
```diff
@@ -167,32 +249,41 @@
...
+        });
   }
 
   /**
    * Delete a document from the database.
    *
    * @param documentReference The DocumentReference to delete.
-   * @return An ApiFuture containing the result of the delete. Contains an error if the delete
-   *     fails.
+   * @return An ApiFuture containing a sentinel value (Timestamp(0)) for the delete operation.
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:273]: Can we explain why the sentinel value is returned? (Also, 0 probably isn't the best sentinel value since Jan 1 1970 is a pretty practical date for a lot of data). This I think also differs from the behavior of `DocRef.delete`and `batch.delete` which I believe use the server commit response time.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 25]
```diff
@@ -201,16 +292,21 @@
...
   public ApiFuture<WriteResult> delete(
-      @Nonnull DocumentReference documentReference, @Nonnull Precondition precondition) {
+      @Nonnull final DocumentReference documentReference,
+      @Nonnull final Precondition precondition) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future = bulkCommitBatch.delete(documentReference, precondition);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:302]: More missing code coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 26]
```diff
@@ -240,18 +341,23 @@
...
-      @Nonnull SetOptions options) {
+      @Nonnull final DocumentReference documentReference,
+      @Nonnull final Map<String, Object> fields,
+      @Nonnull final SetOptions options) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future = bulkCommitBatch.set(documentReference, fields, options);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:353]: Coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 27]
```diff
@@ -262,16 +368,23 @@
...
-      @Nonnull DocumentReference documentReference, Object pojo, @Nonnull SetOptions options) {
+      @Nonnull final DocumentReference documentReference,
+      @Nonnull final Object pojo,
+      @Nonnull final SetOptions options) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future = bulkCommitBatch.set(documentReference, pojo, options);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:380]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 28]
```diff
@@ -280,15 +393,21 @@
...
   @Nonnull
-  public ApiFuture<WriteResult> set(@Nonnull DocumentReference documentReference, Object pojo) {
+  public ApiFuture<WriteResult> set(
+      @Nonnull final DocumentReference documentReference, @Nonnull final Object pojo) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future = bulkCommitBatch.set(documentReference, pojo);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:403]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 29]
```diff
@@ -326,18 +451,23 @@
...
-      Precondition precondition) {
+      @Nonnull final DocumentReference documentReference,
+      @Nonnull final Map<String, Object> fields,
+      final Precondition precondition) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future = bulkCommitBatch.update(documentReference, fields, precondition);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:463]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 30]
```diff
@@ -353,20 +483,24 @@
...
+      @Nonnull final String field,
+      @Nullable final Object value,
+      final Object... moreFieldsAndValues) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future =
-        bulkCommitBatch.update(documentReference, field, value, moreFieldsAndValues);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:496]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 31]
```diff
@@ -382,20 +516,24 @@
...
+      @Nonnull final FieldPath fieldPath,
+      @Nullable final Object value,
+      final Object... moreFieldsAndValues) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future =
-        bulkCommitBatch.update(documentReference, fieldPath, value, moreFieldsAndValues);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:529]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 32]
```diff
@@ -411,21 +549,25 @@
...
+      @Nonnull final String field,
+      @Nullable final Object value,
+      final Object... moreFieldsAndValues) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future =
-        bulkCommitBatch.update(documentReference, precondition, field, value, moreFieldsAndValues);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:563]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 33]
```diff
@@ -442,22 +584,178 @@
...
+      @Nullable final Object value,
+      final Object... moreFieldsAndValues) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future =
-        bulkCommitBatch.update(
-            documentReference, precondition, fieldPath, value, moreFieldsAndValues);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:598]: coverage

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 34]
```diff
@@ -477,28 +775,66 @@
...
             MoreExecutors.directExecutor());
+
     return flushComplete;
   }
 
   /**
    * Commits all enqueued writes and marks the BulkWriter instance as closed.
    *
-   * <p>After calling `close()`, calling any method wil return an error.
+   * <p>After calling `close()`, calling any method wil return an error. Any retries scheduled with
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:836]:
```suggestion
   * <p>After calling `close()`, calling any method will return an error. Any retries scheduled with
```

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 35]
```diff
@@ -477,28 +775,66 @@
...
+   * `addWriteErrorListener()` will be run before the `close()` completes.
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:837]: Is the same true for success listeners? If so let's document it.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterException.java [hunk 36]
```diff
@@ -0,0 +1,68 @@
...
+
+package com.google.cloud.firestore;
+
+import com.google.cloud.firestore.BulkWriter.OperationType;
+import io.grpc.Status;
+
+/** The error thrown when a BulkWriter operation fails. */
+final class BulkWriterException extends FirestoreException {
+  /** @return The status code of the error. */
+  public Status getStatus() {
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterException.java:25]: nit: Java style usually has getters below constructors and constructors below fields.

File: google-cloud-firestore/clirr-ignored-differences.xml [hunk 37]
```diff
@@ -223,6 +223,30 @@
...
+  </difference>
+  <difference>
+    <differenceType>6010</differenceType>
+    <className>com/google/cloud/firestore/UpdateBuilder</className>
+    <field>writes</field>
+  </difference>
+  <difference>
+    <differenceType>7002</differenceType>
+    <className>com/google/cloud/firestore/UpdateBuilder</className>
+    <method>int getMutationsSize()</method>
```
Comment (BenWhitehead) [inline google-cloud-firestore/clirr-ignored-differences.xml:247]: We can't remove this method without a major version release since it's public. because it's public customers could be using already and we won't want to remove it and break them.

We should be able to keep the method public without any issues.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java [hunk 38]
```diff
@@ -680,169 +637,18 @@ private T performDelete(
...
+    return writes;
   }
 
-  void markReadyToSend() {
-    if (state == BatchState.OPEN) {
-      state = BatchState.READY_TO_SEND;
-    }
+  /** Get the number of writes. */
+  @VisibleForTesting
+  int getWriteCount() {
```
Comment (BenWhitehead) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java:651]: We need to keep this method public for binary compatibility as I noted in the clirr file.

After making this public we no longer need the `@VisibleForTesting`.

**Review by BenWhitehead — CHANGES_REQUESTED** [review]

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 29 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:476]: added test.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 28 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:416]: added test.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 39]
```diff
@@ -167,32 +249,41 @@
...
   @Nonnull
   public ApiFuture<WriteResult> create(
-      @Nonnull DocumentReference documentReference, @Nonnull Object pojo) {
+      @Nonnull final DocumentReference documentReference, @Nonnull final Object pojo) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future = bulkCommitBatch.create(documentReference, pojo);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:272]: added test.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 26 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:366]: done.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 27 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:393]: done.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 30 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:509]: done.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 31 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:542]: done.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 32 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:576]: done.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 33 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:611]: done.

**Review by BenWhitehead — APPROVED** [review]
